File: sps/src/lib/api/stuckquery.ts
/**
 * Stuck-query snapshots: the "Running queries information during stuck
 * thread" tables the app logs at every valve detection.
 *
 * NO union shapes on purpose (unlike cpumemstats): MSSQL and PGSQL rows
 * share almost nothing, so each flavor gets its own commands and types.
 * A snapshot = one dump moment = one distinct `timestamp`. PGSQL has no
 * blocking table — the blocking command exists for MSSQL only.
 *
 * Conventions as everywhere: Err serialized to String (rejected promise),
 * camelCase JSON, u64 ms timestamps/durations.
 */

import { invoke } from "@tauri-apps/api/core";

// ---------------------------------------------------------------------------
// Snapshot rollups (one row per dump moment, for the snapshot list)
// ---------------------------------------------------------------------------

export interface MssqlSnapshot {
  /** ms epoch of the dump */
  timestamp: number;
  /** rows in stuckquery_mssql at this timestamp */
  queries: number;
  /** rows with blocked_by != 0 */
  blocked: number;
}

/** One dump moment that logged a "Currently Blocking Query Details" table. */
export interface BlockingSnapshot {
  /** ms epoch of the dump */
  timestamp: number;
  /** distinct head blockers at this timestamp */
  chains: number;
  /** total blocked sessions (rows) at this timestamp */
  sessions: number;
}

export interface PgsqlSnapshot {
  /** ms epoch of the dump */
  timestamp: number;
  /** rows in stuckquery_pgsql at this timestamp */
  queries: number;
  /** rows with waiting = true */
  waiting: number;
  /** rows with state = 'idle in transaction' */
  idleInTxn: number;
}

/**
 * ```rust
 * #[tauri::command]
 * fn stuckquery_mssql_snapshots(state: ...) -> Result<Vec<MssqlSnapshot>, String>
 * #[tauri::command]
 * fn stuckquery_pgsql_snapshots(state: ...) -> Result<Vec<PgsqlSnapshot>, String>
 * ```
 * REQUIREMENTS: one row per distinct timestamp (GROUP BY), ordered by
 * timestamp ascending; empty Vec when the table has no rows (a bundle
 * normally contains only one flavor).
 */
export function stuckqueryMssqlSnapshots(): Promise<MssqlSnapshot[]> {
  return invoke("stuckquery_mssql_snapshots");
}
export function stuckqueryPgsqlSnapshots(): Promise<PgsqlSnapshot[]> {
  return invoke("stuckquery_pgsql_snapshots");
}

/**
 * Blocking snapshots are a SEPARATE command (and separate list rows in the
 * UI) — not a count folded into the mssql snapshot rollup.
 *
 * ```rust
 * #[tauri::command]
 * fn stuckquery_mssql_blocking_snapshots(state: ...) -> Result<Vec<BlockingSnapshot>, String>
 * ```
 * REQUIREMENTS: one row per distinct timestamp in stuckquery_mssql_blocking
 * (GROUP BY: chains = COUNT(DISTINCT head_blocker), sessions = COUNT(*)),
 * ordered by timestamp ascending; empty Vec when nothing ever blocked.
 */
export function stuckqueryMssqlBlockingSnapshots(): Promise<BlockingSnapshot[]> {
  return invoke("stuckquery_mssql_blocking_snapshots");
}

// ---------------------------------------------------------------------------
// Per-snapshot query rows
// ---------------------------------------------------------------------------

/**
 * One row of "Currently Running Queries" (MSSQL). Field names mirror
 * parser::stuckquery::RunningQuery verbatim (camelCased by serde) — the
 * parser is the source of truth, this file follows it. All durations ms.
 *
 * SERIALIZATION REQUIREMENTS on the Rust side:
 *  - status must reach the wire lowercase ("runnable"…, rename_all) —
 *    the badge styling keys off the exact strings;
 *  - waitType must serialize as a PLAIN STRING (WaitType::as_str), not
 *    serde's default externally-tagged enum — Unknown(String) would
 *    otherwise arrive as {"Unknown": "..."}.
 */
export interface MssqlQuery {
  sessionId: number;
  /** runnable | running | rollback | sleeping | background | suspended */
  status: string;
  txnId: number;
  /** blocking session id; 0 = not blocked */
  blockedBy: number;
  waitType: string | null;
  waitResource: string | null;
  waitTimeMs: number;
  cpuTimeMs: number;
  logicalReads: number;
  reads: number;
  writes: number;
  elapsed: number;
  /** the statement excerpt from the table */
  statement: string;
  /** full batch text when logged */
  commandText: string;
  command: string;
  login: string;
  host: string;
  db: string;
  program: string;
  hostProcess: number;
  lastRequestEnd: number;
  loginTime: number;
  openTxn: number;
}

/** One row of "Currently Running Queries" (PGSQL). */
export interface PgsqlQuery {
  pid: number;
  queryTime: number | null;
  txnTime: number | null;
  dbName: string;
  /** active | idle in transaction */
  state: string;
  waiting: boolean;
  query: string;
  /** ms epoch of the last state change */
  stateChange: number;
  applicationName: string | null;
  clientHost: string | null;
  clientPort: number | null;
}

/**
 * ```rust
 * #[tauri::command]
 * fn stuckquery_mssql_queries(timestamp: u64, state: ...) -> Result<Vec<MssqlQuery>, String>
 * #[tauri::command]
 * fn stuckquery_pgsql_queries(timestamp: u64, state: ...) -> Result<Vec<PgsqlQuery>, String>
 * ```
 * REQUIREMENTS: rows WHERE timestamp = $1 exactly (the frontend only asks
 * for timestamps it got from the snapshots command); MSSQL ordered by
 * elapsed descending, PGSQL by query_time descending (nulls last).
 */
export function stuckqueryMssqlQueries(timestamp: number): Promise<MssqlQuery[]> {
  return invoke("stuckquery_mssql_queries", { timestamp });
}
export function stuckqueryPgsqlQueries(timestamp: number): Promise<PgsqlQuery[]> {
  return invoke("stuckquery_pgsql_queries", { timestamp });
}

// ---------------------------------------------------------------------------
// Blocking chain (MSSQL only — PGSQL logs no blocking table)
// ---------------------------------------------------------------------------

/**
 * One row of "Currently Blocking Query Details". Field names mirror
 * parser::stuckquery::BlockingQuery verbatim (camelCased by serde).
 * waitType: same plain-string serialization requirement as MssqlQuery.
 */
export interface MssqlBlockingRow {
  /** session at the root of this chain */
  headBlocker: number;
  sessionId: number;
  txnId: number;
  /** direct blocker of this session */
  blockingSessionId: number;
  waitType: string | null;
  /** ms */
  waitDuration: number;
  waitResource: string | null;
  statementStartOffset: number;
  statementEndOffset: number;
  planHandle: string;
  sqlHandle: string;
  mostRecentSqlHandle: string;
  /** depth in the chain (1 = directly under the head blocker) */
  level: number;
  blockerQueryOrMostRecentQuery: string;
}

/**
 * ```rust
 * #[tauri::command]
 * fn stuckquery_mssql_blocking(timestamp: u64, state: ...) -> Result<Vec<MssqlBlockingRow>, String>
 * ```
 * REQUIREMENTS: rows WHERE timestamp = $1; ordered by head_blocker, then
 * level ascending (the frontend indents by level within each head-blocker
 * group); empty Vec when the snapshot logged no blocking table.
 */
export function stuckqueryMssqlBlocking(timestamp: number): Promise<MssqlBlockingRow[]> {
  return invoke("stuckquery_mssql_blocking", { timestamp });
}

// ---------------------------------------------------------------------------
// Long runners: queries observed across MULTIPLE snapshots
// ---------------------------------------------------------------------------

/**
 * One execution seen in several snapshots (MSSQL identity: the same
 * session_id + txn_id still running = the same execution).
 */
export interface MssqlLongRunner {
  sessionId: number;
  txnId: number;
  statement: string;
  login: string;
  /** distinct snapshots this execution appears in */
  snapshots: number;
  /** ms epoch of the first/last snapshot containing it */
  firstSeen: number;
  lastSeen: number;
  maxElapsedMs: number;
  maxCpuTimeMs: number;
  /** snapshots in which it was blocked (blocked_by != 0) */
  blockedIn: number;
}

/** One execution seen in several snapshots (PGSQL identity: pid + query). */
export interface PgsqlLongRunner {
  pid: number;
  query: string;
  dbName: string;
  snapshots: number;
  firstSeen: number;
  lastSeen: number;
  maxQueryTimeMs: number | null;
  /** snapshots in which state = 'idle in transaction' */
  idleInTxnIn: number;
}

/**
 * ```rust
 * #[tauri::command]
 * fn stuckquery_mssql_longrunning(state: ...) -> Result<Vec<MssqlLongRunner>, String>
 * #[tauri::command]
 * fn stuckquery_pgsql_longrunning(state: ...) -> Result<Vec<PgsqlLongRunner>, String>
 * ```
 * REQUIREMENTS: GROUP BY the identity above; only groups present in MORE
 * THAN ONE distinct timestamp; ordered by snapshots desc, then
 * maxElapsedMs / maxQueryTimeMs desc.
 */
export function stuckqueryMssqlLongrunning(): Promise<MssqlLongRunner[]> {
  return invoke("stuckquery_mssql_longrunning");
}
export function stuckqueryPgsqlLongrunning(): Promise<PgsqlLongRunner[]> {
  return invoke("stuckquery_pgsql_longrunning");
}

File: sps/src-tauri/src/store/stuckquery.rs
use std::str::FromStr;

use duckdb::Connection;
use duckdb::types::Value;

use crate::handlers::types::{
    BlockingSnapshot, MSSQLLongRunningQuery, MSSQLLongRunningTxn, MSSQLSnapshot,
    PGSQLLongRunningQuery, PGSQLSnapshot,
};
use crate::parser::WaitType;
use crate::parser::stuckquery::{BlockingQuery, MSSQLStatus, PGSQLQuery, PGSQLState, RunningQuery};
use crate::store::error::Error;
use crate::store::tables::Tables;

pub fn get_stuckquery_mssql_snapshots(cnx: &Connection) -> Result<Vec<MSSQLSnapshot>, Error> {
    let query = format!(
        "SELECT timestamp, COUNT(session_id), COUNT(blocked_by) FILTER (WHERE blocked_by != 0) FROM {0} GROUP BY timestamp ORDER BY timestamp",
        Tables::StuckqueryMSSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([])?;
    let mut snapshots = Vec::new();
    while let Some(row) = rows.next()? {
        let snapshot = MSSQLSnapshot {
            timestamp: row.get(0)?,
            queries: row.get(1)?,
            blocked: row.get(2)?,
        };
        snapshots.push(snapshot);
    }
    Ok(snapshots)
}

pub fn get_stuckquery_mssql_blocking_snapshots(
    cnx: &Connection,
) -> Result<Vec<BlockingSnapshot>, Error> {
    let query = format!(
        "SELECT timestamp, COUNT(DISTINCT head_blocker), COUNT(session_id) FROM {0} GROUP BY timestamp ORDER BY timestamp",
        Tables::StuckqueryBlockingMSSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([])?;
    let mut snapshots = Vec::new();
    while let Some(row) = rows.next()? {
        let snapshot = BlockingSnapshot {
            timestamp: row.get(0)?,
            chains: row.get(1)?,
            sessions: row.get(2)?,
        };
        snapshots.push(snapshot);
    }
    Ok(snapshots)
}

pub fn get_stuckquery_pgsql_snapshots(cnx: &Connection) -> Result<Vec<PGSQLSnapshot>, Error> {
    let query = format!(
        "SELECT timestamp, COUNT(pid), COUNT(waiting) FILTER (WHERE waiting = true), COUNT(state) FILTER (WHERE state = 'idle in transaction') FROM {0} GROUP BY timestamp ORDER BY timestamp",
        Tables::StuckqueryPGSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([])?;
    let mut snapshots = Vec::new();
    while let Some(row) = rows.next()? {
        let snapshot = PGSQLSnapshot {
            timestamp: row.get(0)?,
            queries: row.get(1)?,
            waiting: row.get(2)?,
            idle_in_txn: row.get(3)?,
        };
        snapshots.push(snapshot);
    }
    Ok(snapshots)
}

pub fn get_stuckquery_pgsql_queries<'a, 'b>(
    cnx: &'a Connection,
    timestamp: u64,
) -> Result<Vec<PGSQLQuery<'b>>, Error> {
    let query = format!(
        "SELECT pid, query_time, txn_time, db_name, state, waiting, query, state_change, application_name, client_host, client_port FROM {} WHERE timestamp = $1",
        Tables::StuckqueryPGSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([timestamp])?;
    let mut queries = Vec::new();
    while let Some(row) = rows.next()? {
        let state: String = row.get(4)?;
        let state = PGSQLState::from_str(&state).unwrap_or(PGSQLState::Active);
        let query = PGSQLQuery {
            pid: row.get(0)?,
            query_time: row.get(1)?,
            txn_time: row.get(2)?,
            db_name: row.get::<usize, String>(3)?.into(),
            state,
            waiting: row.get(5)?,
            query: row.get::<usize, String>(6)?.into(),
            state_change: row.get(7)?,
            application_name: row.get::<usize, Option<String>>(8)?.map(|a| a.into()),
            client_addr: None,
            client_host: row.get::<usize, Option<String>>(9)?.map(|h| h.into()),
            client_port: row.get::<usize, Option<u16>>(10)?,
        };
        queries.push(query);
    }
    Ok(queries)
}

pub fn get_stuckquery_mssql_queries<'a>(
    cnx: &Connection,
    timestamp: u64,
) -> Result<Vec<RunningQuery<'a>>, Error> {
    let query = format!(
        "SELECT session_id, status, txn_id, blocked_by, wait_type, wait_resource, wait_time_ms, cpu_time_ms, logical_reads, reads, writes, elapsed, statement, command_text, command, login, host, db, program, host_process, last_request_end, login_time, open_txn FROM {0} WHERE {0}.timestamp = $1",
        Tables::StuckqueryMSSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([timestamp])?;
    let mut queries = Vec::new();
    while let Some(row) = rows.next()? {
        let status =
            MSSQLStatus::from_str(&row.get::<usize, String>(1)?).unwrap_or(MSSQLStatus::Background);
        let wait_type = row
            .get::<usize, Option<String>>(4)?
            .map(|s| WaitType::parse(&s));
        let query = RunningQuery {
            session_id: row.get(0)?,
            status,
            txn_id: row.get(2)?,
            blocked_by: row.get(3)?,
            wait_type,
            wait_resource: row.get::<usize, Option<String>>(5)?.map(|s| s.into()),
            wait_time_ms: row.get(6)?,
            cpu_time_ms: row.get(7)?,
            logical_reads: row.get(8)?,
            reads: row.get(9)?,
            writes: row.get(10)?,
            elapsed: row.get(11)?,
            statement: row.get::<usize, String>(12)?.into(),
            command_text: row.get::<usize, String>(13)?.into(),
            command: row.get::<usize, String>(14)?.into(),
            login: row.get::<usize, String>(15)?.into(),
            host: row.get::<usize, String>(16)?.into(),
            db: row.get::<usize, String>(17)?.into(),
            program: row.get::<usize, String>(18)?.into(),
            host_process: row.get(19)?,
            last_request_end: row.get(20)?,
            login_time: row.get(21)?,
            open_txn: row.get(22)?,
        };
        queries.push(query);
    }
    Ok(queries)
}

pub fn get_stuckquery_mssql_blocking<'a>(
    cnx: &Connection,
    timestamp: u64,
) -> Result<Vec<BlockingQuery<'a>>, Error> {
    let query = format!(
        "SELECT head_blocker, session_id, txn_id, blocking_session_id, wait_type, wait_duration, wait_resource, statement_start_offset, statement_end_offset, plan_handle, sql_handle, most_recent_sql_handle, level, blocker_query_or_most_recent_query FROM {0} WHERE {0}.timestamp = $1",
        Tables::StuckqueryBlockingMSSQL.into_str()
    );

    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([timestamp])?;
    let mut queries = Vec::new();
    while let Some(row) = rows.next()? {
        let wait_type = row
            .get::<usize, Option<String>>(4)?
            .map(|s| WaitType::parse(&s));
        let query = BlockingQuery {
            head_blocker: row.get(0)?,
            session_id: row.get(1)?,
            txn_id: row.get(2)?,
            blocking_session_id: row.get(3)?,
            wait_type,
            wait_duration: row.get(5)?,
            wait_resource: row.get::<usize, Option<String>>(6)?.map(|s| s.into()),
            statement_start_offset: row.get(7)?,
            statement_end_offset: row.get(8)?,
            plan_handle: row.get::<usize, String>(9)?.into(),
            sql_handle: row.get::<usize, String>(10)?.into(),
            most_recent_sql_handle: row.get::<usize, String>(11)?.into(),
            level: row.get(12)?,
            blocker_query_or_most_recent_query: row.get::<usize, String>(13)?.into(),
        };
        queries.push(query);
    }
    Ok(queries)
}

pub fn get_stuckquery_mssql_long_running(
    cnx: &Connection,
) -> Result<Vec<MSSQLLongRunningQuery>, Error> {
    let query = format!(
        "SELECT session_id, txn_id, statement, login, COUNT(timestamp), MIN(timestamp), MAX(timestamp), MAX(elapsed), MAX(cpu_time_ms), COUNT(blocked_by) FILTER (WHERE blocked_by != 0) FROM {0} GROUP BY session_id, txn_id, statement, login HAVING COUNT(timestamp) > 1 ORDER BY timestamp",
        Tables::StuckqueryMSSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([])?;
    let mut queries = Vec::new();
    while let Some(row) = rows.next()? {
        let query = MSSQLLongRunningQuery {
            session_id: row.get(0)?,
            txn_id: row.get(1)?,
            statement: row.get(2)?,
            login: row.get(3)?,
            snapshots: row.get(4)?,
            first_seen: row.get(5)?,
            last_seen: row.get(6)?,
            max_elapsed: row.get(7)?,
            max_cpu_time_ms: row.get(8)?,
            blocked_in: row.get(9)?,
        };
        queries.push(query);
    }

    Ok(queries)
}

pub fn get_stuckquery_pgsql_long_running(
    cnx: &Connection,
) -> Result<Vec<PGSQLLongRunningQuery>, Error> {
    let query = format!(
        "SELECT pid, query, FIRST(db_name), COUNT(timestamp), MIN(timestamp), MAX(timestamp), MAX(query_time), COUNT(state) FILTER (WHERE state = 'idle in transaction') FROM {} GROUP BY pid, query HAVING COUNT(timestamp) > 1 ORDER BY timestamp",
        Tables::StuckqueryPGSQL.into_str()
    );
    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([])?;
    let mut queries = Vec::new();
    while let Some(row) = rows.next()? {
        let query = PGSQLLongRunningQuery {
            pid: row.get(0)?,
            query: row.get(1)?,
            db_name: row.get(2)?,
            snapshots: row.get(3)?,
            first_seen: row.get(4)?,
            last_seen: row.get(5)?,
            max_query_time_ms: row.get(6)?,
            idle_in_txn_in: row.get(7)?,
        };
        queries.push(query);
    }
    Ok(queries)
}

pub fn get_stuckquery_mssql_long_running_txn(
    cnx: &Connection,
) -> Result<Vec<MSSQLLongRunningTxn>, Error> {
    let query = format!(
        "SELECT session_id, txn_id, FIRST(login), COUNT(timestamps), MIN(timestamps), MAX(timestamps), LIST(statement_text) FROM {} GROUP BY session_id, txn_id ORDER BY timestamp",
        Tables::StuckqueryMSSQL.into_str()
    );

    let mut stmt = cnx.prepare_cached(&query)?;
    let mut rows = stmt.query([])?;
    let mut result = Vec::new();
    while let Some(row) = rows.next()? {
        let queries: Value = row.get(6)?;
        let queries: Vec<String> = match queries {
            Value::List(items) => items
                .into_iter()
                .flat_map(|v| match v {
                    Value::Text(s) => Ok(s),
                    _ => Err(()),
                })
                .collect(),
            _ => Vec::new(),
        };
        let query = MSSQLLongRunningTxn {
            session_id: row.get(0)?,
            txn_id: row.get(1)?,
            login: row.get(2)?,
            snapshots: row.get(3)?,
            first_seen: row.get(4)?,
            last_seen: row.get(5)?,
            queries,
        };
        result.push(query);
    }
    Ok(result)
}
